clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/PersistenceDAO.scala

```diff
@@ -78,6 +83,43 @@ trait PersistenceDAO extends LazyLogging {
       )
       logger.debug(s"Wrote document $document to ${written.toUri}")
     }
+
+  /**
+    * Return the GCS bucket metadata files for `index` upserted on or
+    * after `upsertId`.  Decode the files from JSON as `ClioDocument`s
+    * sorted by `upsertId`.
+    *
+    * @param upsertId of the last known upserted document
+    * @param index  to query against.
+    * @param ec is the implicit ExecutionContext
+    * @tparam D is a ClioDocument type with a JSON Decoder
+    * @return a Future Seq of ClioDocument
+    */
+  def getAllSince[D <: ClioDocument: Decoder](
+    upsertId: UUID,
+    index: ElasticsearchIndex[D]
+  )(implicit ec: ExecutionContext): Future[Seq[D]] =
+    Future {
+      val rootDirectory = rootPath.resolve(index.currentPersistenceDir)
+      val files = Try(Files.newDirectoryStream(rootDirectory))
+        .map(_.asScala)
```

Comment:
Thinking of performance too, `_.asScala` will pull the entire stream into memory (not sure if `Files.newDirectoryStream` is lazy in the first place though). It might be worth trying to use Alpakka's `Directory.ls` connector instead: http://developer.lightbend.com/docs/alpakka/current/file.html#directory

Reply:
`newDirectoryStream()` is a lazy iterator, I think it reads one element ahead at a time.

I'm not sure how to do this in Scala, in java I'd use a single streaming operation, using `filter` to skip all files until you hit the one you want to recover from. Although this only works if the files are in insertion order.

Reply:
We gotta sort something in memory, but I reduced it somewhat.